languages/python/sqlalchemy-oso/sqlalchemy_oso/roles.py

```diff
@@ -1,16 +1,142 @@
+
+
+    :param declarative_base: The SQLAlchemy declarative base model that \
+    the role model and all related models are mapped to.
+
+    :param user_model: The SQLAlchemy model representing users that the \
+    resource-specific roles can be assigned to. The generated Role mixin will \
+    have a many-to-one (Foreign Key) relationship with this user model. \
+    A many-to-many relationship to ``resource_model`` is added to ``user_model``; \
+    the relationship is named following the convention: ``resource_model.__name__.lower() + "s"``.
+
+    :param resource_model: The SQLAlchemy model representing resources that \
+    the generated Role mixin will be scoped to. The Role mixin will \
+    have a many-to-one (ForeignKey) relationship with this resource model. \
+    A many-to-many relationship to ``user_model`` is added to ``resource_model``; \
+    the relationship is named ``users``.
+
+    :param roles: An order-independent list of the built-in roles for this resource-specific role type.
+    :type roles: List[str]
+
+        .. code-block:: python
+
+            class Team(Base):
+                __tablename__ = "teams"
+
+                id = Column(Integer, primary_key=True)
+                name = Column(String(256))
+
+    :param mutually_exclusive: Boolean flag that sets whether or not users \
+    can have more than one role for a given resource. Defaults to ``True``.
+    :type roles: bool
+
+    :return: the ResourceRole mixin, which must then be mixed into a SQLAlchemy model for the role. E.g.,
+
+        .. code-block:: python
+
+            OrganizationRoleMixin = oso_roles.resource_role_class(
+                Base, User, Organization, ["OWNER", "MEMBER", "BILLING"]
+            )
+
+            class OrganizationRole(Base, OrganizationRoleMixin):
+                pass
+
+
+    """
+    global ROLE_CLASSES
+    ROLE_CLASSES.append(
+        {
+            "user_model": user_model,
+            "resource_model": resource_model,
+        }
+    )
+
+    tablename = f"{resource_model.__name__.lower()}_roles"
+    if mutually_exclusive:
+        unique_constraint = UniqueConstraint(
+            f"{resource_model.__name__.lower()}_id", "user_id"
+        )
+    else:
+        unique_constraint = UniqueConstraint(
+            f"{resource_model.__name__.lower()}_id", "name", "user_id"
+        )
+
+    class ResourceRoleMixin:
+        choices = role_choices
+
+        __tablename__ = tablename
+        id = Column(Integer, primary_key=True)
+        name = Column(String())
+        __table_args__ = (unique_constraint,)
+
+        @validates("name")
+        def validate_name(self, key, name):
+            if name not in self.choices:
+                raise ValueError(
+                    f"{name} Is not a valid choice for {self.__class__.__name__}"
+                )
+            return name
+
+        @declared_attr
+        def user_id(self):
+            type = inspect(user_model).primary_key[0].type
```

Comment:
Nit: would this recompute this every time user_id is called? Would it be better to compute this outside the method definition?

Reply:
(Ditto for below)

Reply:
Oh nevermind, I think this is only computed once on defining the class

Reply:
Yes I think that's right. To make it more clear I'll change `self` to `cls`